File: TAREA 3/2_DADOS_2_ENTEROS_POSITIVOS/Combinacion.py
# combinacion.py

def combinacion(n: int, k: int) -> int:
    """
    Calcula la combinación C(n, k) = n! / (k! (n-k)!)
    usando una forma multiplicativa eficiente.
    """
    if n < 0 or k < 0:
        raise ValueError("n y k deben ser enteros positivos")

    if k > n:
        return 0

    # Usar simetría para reducir operaciones: C(n,k)=C(n,n-k)
    k = min(k, n - k)

    if k == 0:
        return 1

    numer = 1  # parte de arriba
    denom = 1  # parte de abajo (k!)
    
    for i in range(1, k + 1):
        numer *= n - (k - i)
        denom *= i

    return numer // denom  # división exacta


def main():
    print("CÁLCULO DE COMBINACIÓN C(n,k)\n")

    n = int(input("Ingrese n (entero positivo): "))
    k = int(input("Ingrese k (entero positivo): "))

    resultado = combinacion(n, k)
    print(f"\nC({n}, {k}) = {resultado}")


if __name__

File: TAREA 3/2_DADOS_2_ENTEROS_POSITIVOS/Combinacion.java
// Combinacion.java
import java.math.BigInteger;
import java.util.Scanner;

public class Combinacion {

    public static BigInteger combinacion(int n, int k) {
        if (n < 0 || k < 0)
            throw new IllegalArgumentException("n y k deben ser positivos");

        if (k > n)
            return BigInteger.ZERO;

        k = Math.min(k, n - k); // C(n,k) == C(n,n-k)

        if (k == 0)
            return BigInteger.ONE;

        BigInteger numer = BigInteger.ONE;
        BigInteger denom = BigInteger.ONE;

        for (int i = 1; i <= k; i++) {
            numer = numer.multiply(BigInteger.valueOf(n - (k - i)));
            denom = denom.multiply(BigInteger.valueOf(i));
        }

        return numer.divide(denom);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("CÁLCULO DE COMBINACIÓN C(n,k)");
        System.out.print("Ingrese n: ");
        int n = sc.nextInt();
        System.out.print("Ingrese k: ");
        int k = sc.nextInt();

        BigInteger resultado = combinacion(n, k);

        System.out.println("\nC(" + n + ", " + k + ") = " + resultado);
        sc.close();
    }
}
